### ISSUE TITLE
[Bug][move-compiler-v2] ExpRewriter is error-prone on `Assign` statements

### ISSUE THREAD

**Issue by brmataptos** [aptos-labs/aptos-core#12442]
# 🐛 Bug

Recently the following check was removed from `impl<'env, 'rewriter> ExpRewriterFunctions for ExpRewriter<'env, 'rewriter>` in `third_party/move/move-model/src/exp_rewriter.rs`:
```
     fn rewrite_pattern(&mut self, _pat: &Pattern, entering_scope: bool) -> Option<Pattern> {
-        if !entering_scope {
-            panic!("ExpRewriter won't work correctly on code with `Assign` operation.");
-        }
         None
     }
```
I had added this check here because (1) `ExpRewriter` is used to do simple renaming of free variables in an expression, using
a simple function `replacer:  &'rewriter mut F` which rewrites occurrences of `LocalVar` and Temporary` expressions with free variables. (2) Free variables can also occur in `Assign` expressions, where they appear not as expressions but as `Pattern::Var` enum instances.

The prover never uses the `ExpRewriter` on any code with `Assign` statements, so this assertion could never trigger.  But the Lambda Lifting pass uses the same `ExpRewriter` to rewrite free variables in a `Lambda`.  In the case of Lambda Lifting, it is ok ignore variables in `Assign` operations, as a separate pass checks for symbols in `Assign`: free variables whose symbols appear unshadowed in `Assign` are marked as modified, generating a user error, so it is safe to ignore `Assign` patterns.  So the simple assertion is violated in this case but the code is still safe, so the assertion was removed.

However, it's too easy to use `ExpRewriter` in a situation where it does the wrong thing with `Assign`.  In fact, significant time was wasted during implementation of `Inlining` because `ExpRewriter` seemed to be applicable.  Only thorough testing and debugging caught the problem.

We still need a mechanism to keep others from wandering into this briar patch.

**Comment by brmataptos** [comment]
One simple option is to rename `ExpRewriter` to some more specific name that conveys the limited applicability: `ExpRewriterIgnoringAssign` has been rejected by @wrwg but perhaps there is a more graceful alternative name.  In any case, a prominent warning is needed in a comment at least.

**Comment by wrwg** [comment]
The solution here is most likely to pass an extra parameter into the pattern rewriter, indicating whether the pattern is used in defining (Let) or referencing (Assign) mode. In fact that is exactly how the type checker deals with patterns.

**Comment by brmataptos** [comment]
Fixed by #11553
